pandas/core/frame.py

```diff
@@ -5512,6 +5513,82 @@ def update(
 
     # ----------------------------------------------------------------------
     # Data reshaping
+    @Appender(
+        """
+Examples
+--------
+>>> df = pd.DataFrame({'Animal': ['Falcon', 'Falcon',
+...                               'Parrot', 'Parrot'],
+...                    'Max Speed': [380., 370., 24., 26.]})
+>>> df
+   Animal  Max Speed
+0  Falcon      380.0
+1  Falcon      370.0
+2  Parrot       24.0
+3  Parrot       26.0
+>>> df.groupby(['Animal']).mean()
+        Max Speed
+Animal
+Falcon      375.0
+Parrot       25.0
+
+**Hierarchical Indexes**
+
+We can groupby different levels of a hierarchical index
+using the `level` parameter:
+
+>>> arrays = [['Falcon', 'Falcon', 'Parrot', 'Parrot'],
+...           ['Captive', 'Wild', 'Captive', 'Wild']]
+>>> index = pd.MultiIndex.from_arrays(arrays, names=('Animal', 'Type'))
+>>> df = pd.DataFrame({'Max Speed': [390., 350., 30., 20.]},
+...                   index=index)
+>>> df
+                Max Speed
+Animal Type
+Falcon Captive      390.0
+       Wild         350.0
+Parrot Captive       30.0
+       Wild          20.0
+>>> df.groupby(level=0).mean()
+        Max Speed
+Animal
+Falcon      370.0
+Parrot       25.0
+>>> df.groupby(level="Type").mean()
+         Max Speed
+Type
+Captive      210.0
+Wild         185.0
+"""
+    )
+    @Appender(_shared_docs["groupby"] % _shared_doc_kwargs)
+    def groupby(
+        self,
+        by=None,
+        axis=0,
+        level=None,
+        as_index: bool = True,
+        sort: bool = True,
+        group_keys: bool = True,
+        squeeze: bool = False,
+        observed: bool = False,
+    ) -> "grp_generic.DataFrameGroupBy":
```

Comment:
Could you do ``from pandas.core.groupby.generic import DataFrameGroupBy``? That would make the signature a bit cleaner for end users ("grp_generic" is meaningless for them)

Reply:
That's not possible, because it would give circular import errors.

Maybe in the future, when we we're om min python 3.8, we can do ``from __future__ import annotations`` and get it to work, but that's a long way in the future.

Reply:
Also not if you put it behind a `if TYPE_CHECKING:` ?

Reply:
I actually also don't understand this aspect of importing: what's the difference between `from pandas.core.groupby import generic` vs `from pandas.core.groupby.generic import DataFrameGroupBy`? Both import the `pandas.core.groupby.generic` (and thus run all imports in that file), no?

Reply:
I'm not sure I can articulate it, but I understand it like this: If you attach a name to the new module (like you'd do in ``from pandas.core.groupby.generic import DataFrameGroupBy``) you force the ``pandas.core.groupby.generic`` module to be run at import time. This causes an error, if the calling module (here ``frame.py``) itself must be run, before ``groupby.generic``, casing a circular error.

By not importing a name (here ``DataFrameGroupby``), running the module is delayed, and things work.